pvlib/iotools/epw.py

```diff
@@ -0,0 +1,181 @@
+    liquid_precipitation_depth          The amount of liquid precipitation observed at indicated time for the period indicated in the liquid precipitation quantity field, millimeter
+    liquid_precipitation_quantity       The period of accumulation for the liquid precipitation depth field, hour
+    =============================       ======================================================================================================================================================
+
+    References
+    ----------
+
+    [1] EnergyPlus documentation, Auxiliary Programs
+    https://energyplus.net/documentation.
+    '''
+    
+    if filename.startswith('http'):
+        # Attempts to download online EPW file
+        # See comments above for possible online sources
+        request = Request(filename, headers={'User-Agent': (
+            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_5) '
+            'AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.87 '
+            'Safari/537.36')})
+        response = urlopen(request)
+        csvdata = io.StringIO(response.read().decode(errors='ignore'))
+    else:
+        # Assume it's accessible via the file system
+        csvdata = open(filename, 'r')
+    
+    # Read line with metadata
+    firstline = csvdata.readline()
+
+    head = ['loc','city', 'state-prov', 'country', 'data_type','WMO_code', 
+            'latitude', 'longitude', 'TZ','altitude']
+    meta = dict(zip(head, firstline.rstrip('\n').split(",")))
+
+    meta['altitude'] = float(meta['altitude'])
+    meta['latitude'] = float(meta['latitude'])
+    meta['longitude'] = float(meta['longitude'])
+    meta['TZ'] = float(meta['TZ'])
+
+    colnames = ['year', 'month', 'day', 'hour', 'minute', 'data_source_unct', 
+                'temp_air', 'temp_dew', 'relative_humidity', 
+                'atmospheric_pressure', 'etr', 'etrn', 'ghi_infrared', 'ghi', 
+                'dni', 'dhi', 'global_hor_illum', 'direct_normal_illum', 
+                'diffuse_horizontal_illum', 'zenith_luminance', 
+                'wind_direction', 'wind_speed', 'total_sky_cover', 
+                'opaque_sky_cover', 'visibility', 'ceiling_height', 
+                'present_weather_observation', 'present_weather_codes',
+                'precipitable_water', 'aerosol_optical_depth', 'snow_depth', 
+                'days_since_last_snowfall', 'albedo', 
+                'liquid_precipitation_depth', 'liquid_precipitation_quantity']
+   
+    # We only have to skip 6 rows instead of 7 because we have already used 
+    # the realine call above.
+    data = pd.read_csv(csvdata, skiprows=6, header=0, names=colnames)
+    
+    # Change to single year if requested
+    if coerce_year is not None:
+        data["year"] = coerce_year
+        data['year'].iloc[-1] = coerce_year - 1
+
+    # Update index with correct date information     
+    data = data.set_index(pd.to_datetime(data[['year', 'month', 'day', 
```

Comment:
This line causes problems on the python 2.7-min configuration. Have you checked that the index is correct in your debugging? It's not explicitly tested in any of the tests. https://travis-ci.org/pvlib/pvlib-python/jobs/510118505#L914

Reply:
Apologies if this is a silly question, but I think I need a bit more explanation.
What would 'correctness' of the index mean?
Do you have any example of how this can be tested?

Reply:
I only meant that the index is what you expected it to be. Here's a more formal example for the index expectation for the [read_solrad test function](https://github.com/pvlib/pvlib-python/blob/master/pvlib/test/test_solrad.py#L23-L27).

Reply:
Suggest creating the index separately, and localizing it, then setting the dataframe index.
E.g.,

`idx = pd.DatetimeIndex(pd.to_datetime(data[['year', 'month', 'day', 'hour']])`
`idx.tz_localize(xxx)`
`data.index = idx`

Reply:
I managed to follow Cliff's suggestion for separately creating the index first. This seems to work fine.
But I am getting pretty much stuck with the error in the python 2.7-min configuration.
I have verified that the index gets created, and that it makes sense. The zero padding solution is not helping.
I think that the issue is that the index comes in the form of a series, which `to_datetime` does not like, apparently.
I have tried to solutions proposed [here](https://stackoverflow.com/questions/46231003/attributeerror-dataframe-object-has-no-attribute-dtype-using-pd-to-dateti), but it is not working for me.
Any ideas or suggestions would be very welcome.

Reply:
The problem is related to the time specification using hour = 24. I checked out your code and got it to work like this:

```python
    # create index that supplies correct date and time zone information
    dts = data[['month', 'day']].astype(str).apply(lambda x: x.str.zfill(2))
    hrs = (data['hour'] - 1).astype(str).str.zfill(2)
    dtscat = data['year'].astype(str) + dts['month'] + dts['day'] + hrs
    idx = pd.to_datetime(dtscat, format='%Y%m%d%H')
    idx = idx.dt.tz_localize(int(meta['TZ'] * 3600))
    data.index = idx
```

Consider adding a +1 hour timedelta to the index. In any case, we'll need to make sure the docstring correctly describes the index.

Reply:
@roelloonen , @cwhanse, @janinefreeman, @wholmgren :  Hi all- I'm curious about the -1 hour offset in read_epw() in the current pvlib release (line 171).  I think that this is a mistake - when I try to match modeled dni using ghi and dhi, I can only do this when I  calculate solar position at (timestamp + 30 minutes), which is different from the TMY convention of calculating solar position at (timestamp - 30 minutes).

Reply:
@cdeline The shift is because EPW files follow an hour-ending timestamp convention, and thus hours take values from 1 to 24.  Hour values are shifted to hour-beginning to avoid changing the date for the hour=24 timestamp in the pandas index.

The docstring for `read_epw` describes hour-ending convention for all variables, which needs to be corrected since we've shifted to hour beginning.

I'm confused about the line number you cite: I see that shift at line [165 ](https://github.com/pvlib/pvlib-python/blob/a72aca77a071937ccd740d5f0c6b570f030b6538/pvlib/iotools/epw.py#L165) of `read_epw` - is it numbered 171 in your IDE, or are we actually looking at different code?

Reply:
Cliff-  I was referring to V0.6.3 which has an additional bit of importing at the top:
try:
    # python 2 compatibility
    from urllib2 import urlopen, Request
except ImportError:
    from urllib.request import urlopen, Request

Otherwise, I think it's identical to this 0.6.1 version here.  Thanks for your discussion above.